statically linked gnutls, fixed windows build, no more -dirty

iameli opened the pull request:
Three things here:

* statically-linked gnutls on Linux and Mac, no more external dependency (#1094)
* tweaks to the Windows build to get Go 1.13 and gomodules working (#1121)
* fix the `-dirty` suffix showing up on the Linux build by preinstalling the necessary go linting and testing tools prior to the test execution. (i never actually filed this but this commit incorporates #1095)

The Windows build still has some external DLL dependencies but Windows apps are pretty routinely distributed with an `.exe` and its supporting `.dll`s so I don't see that as a big deal.

Fixes #1094 #1121

Proof:

```
> ldd ./livepeer
	linux-vdso.so.1 (0x00007fff109a7000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f9559933000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f955992e000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f95597ab000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f955958d000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f9559583000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f95593c2000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f9559959000)
```

```
> otool -L ./livepeer                                                                                            19:06:31
./livepeer:
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 1570.15.0)
	/System/Library/Frameworks/Security.framework/Versions/A/Security (compatibility version 1.0.0, current version 58286.251.4)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/System/Library/Frameworks/VideoToolbox.framework/Versions/A/VideoToolbox (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/CoreMedia.framework/Versions/A/CoreMedia (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/CoreVideo.framework/Versions/A/CoreVideo (compatibility version 1.2.0, current version 1.5.0)
	/System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices (compatibility version 1.0.0, current version 944.3.0)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
```

iameli committed:
ci: prevent upload if git state is dirty

mk-livepeer committed:
Makefile: `go build -mod=readonly` to avoid dirty git state

mk-livepeer committed:
git diff && git status when git state dirty

iameli commented:
Okay, ready for review. Let me know if you'd like me to clean up this big ol' pile of commits into something cleaner. @darkdarkdragon @j0sh

j0sh commented on install_ffmpeg.sh:
```diff
@@ -35,7 +36,7 @@ fi
 if [ ! -e "$HOME/nasm/nasm" ]; then
   # sudo apt-get -y install asciidoc xmlto # this fails :(
-  git clone -b nasm-2.14.02 https://repo.or.cz/nasm.git "$HOME/nasm"
+  git clone -b nasm-2.14.02 https://github.com/iameli/nasm.git "$HOME/nasm"
```
Is this needed?

iameli replied:
Oops, no - `repo.or.cz` was briefly down while I was hacking on this. Switched back.

j0sh commented on install_ffmpeg.sh:
```diff
@@ -53,7 +54,52 @@ if [ ! -e "$HOME/x264" ]; then
   make install-lib-static
 fi
 
+# Static linking of gnutls on Linux/Mac
```
We might not need to build gnutls on Mac if the ffmpeg build can be configured to use the built-in securetransport library. That should be auto-detetected under normal circumstances, but `--enable-securetransport` should work to make it explicit.

iameli replied:
Oh nice! I'll try running a build like that and make sure it works - I had trouble before with cert errors when I was trying to have mbedtls and gnutls talk to each other, but that could totally work yeah.

j0sh replied:
It's quite possible that how we generate certs needs fixing as well. Last I tried, curl could not query O status, even with `--no-check-certificate`. So it'd be nice if the certs weren't *completely* bogus, but that may be a separate issue entirely.

iameli replied:
Hmmm, odd. I took out `--enable-gnutls` on the macOS builds. It looked to have the desired effect on FFMpeg's `./configure`, it changed from

```
External libraries:
appkit			coreimage		zlib
avfoundation		gnutls
bzlib			libx264
```

to

```
External libraries:
appkit			coreimage		zlib
avfoundation		libx264
bzlib			securetransport
```

But then, when I try to run the binary with gnutls uninstalled...

```
▶ ./livepeer-darwin-amd64/livepeer -version                                      17:27:20
dyld: Library not loaded: /usr/local/opt/gnutls/lib/libgnutls.30.dylib
  Referenced from: /Users/iameli/code/go/src/github.com/livepeer/go-livepeer/./livepeer-darwin-amd64/livepeer
  Reason: image not found
fish: './livepeer-darwin-amd64/livepee…' terminated by signal SIGABRT (Abort)
```

Something else in the dependencies must be linking it, I suppose? Gonna try explicitly uninstalling it prior to the build.

iameli replied:
Ah, here's the trouble: https://github.com/livepeer/lpms/blob/master/ffmpeg/ffmpeg.go#L14

I'm sort of inclined to keep gnutls on all the platforms? - It works great and we eliminate any inconsistencies that would occur like I saw with mbedtls

j0sh replied:
> I'm sort of inclined to keep gnutls on all the platforms?

Ah gotcha - yeah, makes sense to do that if it's most expedient.

j0sh commented:
> you'd like me to clean up this big ol' pile of commits into something cleaner.

Would be good to fixup them, and if it's possible, split into separate commits for the gnutls linking, the windows build fix, and the -dirty fix. `git rebase -i` in order to fixup into a single commit, then `git reset HEAD^` followed by `git add -p` to split out the parts specific to each fix.

iameli commented:
Cool, will do the rebase after the builds pass

iameli committed:
linux build: fix -dirty suffix

this preloads the linting and testing tools into the Docker container
so they don't get added to go.mod in the CI process and cause the
-dirty builds to be uploaded

iameli committed:
windows build: add go 1.13 and gomod on windows

iameli committed:
mac and linux builds: add static gnutls

this eliminates the runtime dependency on gnutls on these platforms

mkrufky commented on .circleci/config.yml:
```diff
@@ -12,7 +12,7 @@ workflows:
   build-builder:
     docker:
-      - image: circleci/golang:1-stretch
+      - image: circleci/golang:1.13-stretch
```
I *think* `golang:1-stretch` will present the current stable version (1.13 today, and 1.14 whenever that becomes stable).  If this got inherited from my test PR from a few weeks ago, I was just experimenting there.

mkrufky replied:
changeset ee45869 (from me) can be dropped / reverted

mkrufky commented on docker/Dockerfile.debian:
```diff
@@ -2,7 +2,7 @@
 FROM livepeer/ffmpeg-base:latest as builder
 
-FROM golang:1-stretch as builder2
+FROM golang:1.13-stretch as builder2
```
As I commented earlier, I *think* `golang:1-stretch` will present the current stable version (1.13 today, and 1.14 whenever that becomes stable).  If this got inherited from my test PR from a few weeks ago, I was just experimenting there.

mkrufky replied:
changeset fba3bff (from me) can be dropped / reverted

iameli commented:
Dropped the two go 1.13 Docker commits as mentioned by @mk-livepeer
